packages/app/src/mixins/audit.ts

```diff
@@ -0,0 +1,195 @@
+import { FieldType, VaultItem, Field, AuditResult, AuditResultType } from "@padloc/core/src/item";
+import { Vault } from "@padloc/core/src/vault";
+import { LitElement } from "lit";
+import { getCryptoProvider } from "@padloc/core/src/platform";
+import { HashParams } from "@padloc/core/src/crypto";
+import { stringToBytes, bytesToHex } from "@padloc/core/src/encoding";
+import { sub } from "date-fns";
+
+import { app } from "../globals";
+import { passwordStrength } from "../lib/util";
+import { ListItem } from "../elements/items-list";
+
+type Constructor<T> = new (...args: any[]) => T;
+
+export const AuditMixin = <T extends Constructor<LitElement>>(baseElement: T) => {
+    abstract class M extends baseElement {
+        async auditVaults(
+            vaults: Vault[],
+            {
+                updateOnlyItemWithId,
+                updateOnlyIfOutdated,
+            }: { updateOnlyItemWithId?: string; updateOnlyIfOutdated?: boolean } = {}
+        ) {
+            const reusedPasswords: ListItem[] = [];
+            const weakPasswords: ListItem[] = [];
+            const compromisedPasswords: ListItem[] = [];
+
+            // Don't try to run if the app has locked
+            if (app.state.locked) {
+                return {
+                    reusedPasswords,
+                    weakPasswords,
+                    compromisedPasswords,
+                };
+            }
+
+            const firstMatchPerPasswordHash = new Map<
+                string,
+                { item: VaultItem; vault: Vault; passwordField: { field: Field; fieldIndex: number } }
+            >();
+            const reusedPasswordItemIds: Set<string> = new Set();
+            const weakPasswordItemIds: Set<string> = new Set();
+            const compromisedPasswordItemIds: Set<string> = new Set();
+
+            const oneWeekAgo = sub(new Date(), { weeks: 1 });
+
+            // TODO: Remove this
+            console.log(`Running audit! ${JSON.stringify({ updateOnlyItemWithId, updateOnlyIfOutdated })}`);
```

Comment:
Will remove this before merging, just kept it in so you could see how frequently it ran.